pkg/controller/base/reconciler.go

```diff
@@ -0,0 +1,94 @@
+package base
+
+import (
+	"context"
+	"fmt"
+
+	"github.com/go-logr/logr"
+	consts "github.com/infinispan/infinispan-operator/pkg/controller/constants"
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/apimachinery/pkg/labels"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/types"
+	"k8s.io/client-go/tools/record"
+	"reflect"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	logf "sigs.k8s.io/controller-runtime/pkg/log"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+)
+
+const (
+	EventReasonResourceNotReady = "ResourceNotReady"
+)
+
+type ReconcilerBase struct {
```

Comment:
`ReconcilerBase` is an interesting idea, however I'm a little worried by two things.

1. It feels like we're trying to use a inheritance based approach by using a "base" struct, wheras we should be favouring composition.
2. `ReconcilerBase` functions cover a lot of unrelated functionality. Logging, resource lookup, pod labels.

IMO a better approach would be to create a interface that's dedicated to Logs & events. This interface can then be passed as the args to our functions used throughout our code base. For example, `LookupResource` is a generic kubernetes operation so it probably makes more sense in `pkg/kubernetes/controllerutil.go` as it's not controller specific. We then update it's parameters to take the new logs/events interface instead of `logr.Logger` so that we can log and send the event.

Having a dedicated logs/events interace would also help #690.

Reply:
Yes, I agree that combining the `LookupResource` and mixing other functions inside the `ReconcileBase` may have a kind of confusion of concepts. I would keep this structure as the basis for all our controllers and consolidate all the required structures for all potential controllers to avoid redundant declaration and initialization every time we need it, i.e.
```
type ReconcilerBase struct {
	Name string
	client.Client
	kubernetes.Kubernetes
	// TBD some log and event recorder interface or separate interfaces
}
```
Also other `*Label` related common methods should be moved as static go file inside `kubernetes` package like pods.go as they are not API server related
I would like to re-implement common `UpdateResource` functions based on `infinispan_operator.update` inside `kubernetes` package to use `controllerutil.CreateOrPatch` in more shared way for update Infinispan Group CR's (Batch, Cache, Infinispan, etc.)
`LookupResource` and others functions which will require to access log subsystem will need to accept a new log and event recorder interface or separate construction